File: services/shorts/internal/services/shorts/register_review.go
package shorts

// RegisterReviewService — the operator console behind the Register of Interests.
//
// NOTHING HERE IS A PUBLIC READ PATH. Every method is VISIBILITY_PRIVATE with
// required_role="admin" in the proto AND listed in internalOnlyMethods, so it
// needs the internal service secret as well as an admin identity. Declaring a
// method PRIVATE alone is not enough: the interceptor treats PRIVATE with no
// required_role as "any authenticated user".
//
// WHY THE CONSOLE EXISTS. §6.1's item-1 gate is 51.18%, and §8.19.1 showed the
// denominator is a DEFAULT bucket — the resolver labels anything it cannot
// otherwise explain 'listed'. So the backlog contains two different things that
// look identical in the number: names we could match and have not, and names
// that are correctly unmatchable (delisted, foreign, unlisted funds, prose). A
// human separating them raises coverage honestly. A machine doing it would be
// guessing about named individuals, which is the one thing this subsystem does
// not do.

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"connectrpc.com/connect"
	registerreviewv1 "github.com/castlemilk/shorted.com.au/services/gen/proto/go/registerreview/v1"
	shortsstore "github.com/castlemilk/shorted.com.au/services/shorts/internal/store/shorts"
)

// decisionWire maps the closed proto enum onto the alias table's vocabulary.
// The enum has no fuzzy member, so there is no wire value that could ask for an
// unpublishable match — the omission is the enforcement.
var decisionWire = map[registerreviewv1.SecurityDecision]string{
	registerreviewv1.SecurityDecision_SECURITY_DECISION_RESOLVED:       "resolved",
	registerreviewv1.SecurityDecision_SECURITY_DECISION_NOT_A_SECURITY: "not_a_security",
	registerreviewv1.SecurityDecision_SECURITY_DECISION_UNLISTED_FUND:  "unlisted_fund",
	registerreviewv1.SecurityDecision_SECURITY_DECISION_FOREIGN:        "foreign",
	registerreviewv1.SecurityDecision_SECURITY_DECISION_SKIP:           "skip",
}

var aliasKindWire = map[registerreviewv1.AliasKind]string{
	registerreviewv1.AliasKind_ALIAS_KIND_EQUITY:       "equity",
	registerreviewv1.AliasKind_ALIAS_KIND_ETF:          "etf",
	registerreviewv1.AliasKind_ALIAS_KIND_LIC:          "lic",
	registerreviewv1.AliasKind_ALIAS_KIND_MANAGED_FUND: "managed_fund",
}

func (s *ShortsServer) ListSecurityQueue(
	ctx context.Context,
	req *connect.Request[registerreviewv1.ListSecurityQueueRequest],
) (*connect.Response[registerreviewv1.ListSecurityQueueResponse], error) {
	rows, totalCandidates, totalRows, err := s.store.ListSecurityReviewQueue(
		req.Msg.GetLimit(), req.Msg.GetOffset(), req.Msg.GetGateOnly())
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	items := make([]*registerreviewv1.SecurityQueueItem, 0, len(rows))
	for _, r := range rows {
		items = append(items, queueItemToProto(r))
	}
	return connect.NewResponse(&registerreviewv1.ListSecurityQueueResponse{
		Items:           items,
		TotalCandidates: totalCandidates,
		TotalRows:       totalRows,
	}), nil
}

func queueItemToProto(r *shortsstore.SecurityQueueRow) *registerreviewv1.SecurityQueueItem {
	item := &registerreviewv1.SecurityQueueItem{
		CandidateNorm: r.CandidateNorm,
		Example:       r.Example,
		Occurrences:   r.Occurrences,
		People:        r.People,
		Items:         r.Items,
		Parliaments:   r.Parliaments,
		EntityKinds:   r.EntityKinds,
		GateRows:      r.GateRows,
		SkipCount:     r.SkipCount,
	}
	for _, sm := range r.Samples {
		item.Samples = append(item.Samples, &registerreviewv1.DeclaredSample{
			DeclaredText:   sm.DeclaredText,
			PoliticianName: sm.PoliticianName,
			PoliticianSlug: sm.PoliticianSlug,
			ItemNo:         sm.ItemNo,
			Parliament:     sm.Parliament,
			SourceUrl:      sm.SourceURL,
		})
	}
	if r.Proposal != nil {
		p := &registerreviewv1.AliasProposal{
			ProposedStockCode:   r.Proposal.ProposedStockCode,
			ProposedCompanyName: r.Proposal.ProposedCompanyName,
			Confidence:          r.Proposal.Confidence,
			Rationale:           r.Proposal.Rationale,
			Model:               r.Proposal.Model,
			Status:              r.Proposal.Status,
		}
		for _, l := range r.Proposal.Shortlist {
			p.Shortlist = append(p.Shortlist, listingToProto(l))
		}
		item.Proposal = p
	}
	return item
}

func listingToProto(l *shortsstore.RegisterListingRow) *registerreviewv1.Listing {
	return &registerreviewv1.Listing{
		StockCode:        l.StockCode,
		CompanyName:      l.CompanyName,
		ExistingUses:     l.ExistingUses,
		IsStopwordTicker: l.IsStopword,
	}
}

func (s *ShortsServer) SearchListings(
	ctx context.Context,
	req *connect.Request[registerreviewv1.SearchListingsRequest],
) (*connect.Response[registerreviewv1.SearchListingsResponse], error) {
	rows, err := s.store.SearchRegisterListings(req.Msg.GetQuery(), req.Msg.GetLimit())
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	out := make([]*registerreviewv1.Listing, 0, len(rows))
	for _, l := range rows {
		out = append(out, listingToProto(l))
	}
	return connect.NewResponse(&registerreviewv1.SearchListingsResponse{Listings: out}), nil
}

func (s *ShortsServer) DecideSecurityCandidate(
	ctx context.Context,
	req *connect.Request[registerreviewv1.DecideSecurityCandidateRequest],
) (*connect.Response[registerreviewv1.DecideSecurityCandidateResponse], error) {
	decision, ok := decisionWire[req.Msg.GetDecision()]
	if !ok {
		return nil, connect.NewError(connect.CodeInvalidArgument,
			fmt.Errorf("decision is required and must be one of the SecurityDecision values"))
	}

	// THE REVIEWER'S IDENTITY COMES FROM THE REQUEST CONTEXT, never from the
	// body. curated_by is the evidence a human made this call, and a body-supplied
	// identity is an identity the caller chose for themselves — the same shape as
	// the Stripe webhook self-grant the 2026-07 audit found.
	reviewer := reviewerFromHeaders(req.Header())
	if reviewer == "" {
		return nil, connect.NewError(connect.CodeUnauthenticated,
			errors.New("no reviewer identity on the request"))
	}

	affected, err := s.store.DecideSecurityCandidate(
		req.Msg.GetCandidateNorm(), decision, req.Msg.GetStockCode(),
		aliasKindWire[req.Msg.GetAliasKind()], req.Msg.GetNote(), reviewer,
		req.Msg.GetStopwordConfirmed())
	if err != nil {
		switch {
		case errors.Is(err, shortsstore.ErrRegisterListingUnknown),
			errors.Is(err, shortsstore.ErrRegisterStopwordUnconfirmed):
			return nil, connect.NewError(connect.CodeInvalidArgument, err)
		default:
			return nil, connect.NewError(connect.CodeInternal, err)
		}
	}

	coverage, err := s.store.GetRegisterCoverageStats()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(&registerreviewv1.DecideSecurityCandidateResponse{
		RowsAffected: affected,
		Coverage:     coverageToProto(coverage),
	}), nil
}

func (s *ShortsServer) UndoSecurityDecision(
	ctx context.Context,
	req *connect.Request[registerreviewv1.UndoSecurityDecisionRequest],
) (*connect.Response[registerreviewv1.UndoSecurityDecisionResponse], error) {
	deleted, err := s.store.UndoSecurityDecision(req.Msg.GetCandidateNorm())
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	coverage, err := s.store.GetRegisterCoverageStats()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(&registerreviewv1.UndoSecurityDecisionResponse{
		Deleted:  deleted,
		Coverage: coverageToProto(coverage),
	}), nil
}

func (s *ShortsServer) GetCoverageStats(
	ctx context.Context,
	req *connect.Request[registerreviewv1.GetCoverageStatsRequest],
) (*connect.Response[registerreviewv1.GetCoverageStatsResponse], error) {
	coverage, err := s.store.GetRegisterCoverageStats()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(&registerreviewv1.GetCoverageStatsResponse{
		Coverage: coverageToProto(coverage),
	}), nil
}

func coverageToProto(r *shortsstore.RegisterCoverageRow) *registerreviewv1.CoverageStats {
	if r == nil {
		return nil
	}
	return &registerreviewv1.CoverageStats{
		Resolved:          r.Resolved,
		ListedCandidates:  r.ListedCandidates,
		GatePercent:       r.GatePercent,
		BacklogCandidates: r.BacklogCandidates,
		BacklogRows:       r.BacklogRows,
		ClassifiedOut:     r.ClassifiedOut,
		Method:            r.Method,
	}
}

// reviewerFromHeaders reads the admin identity the web tier attached after its
// own requireAdmin() check. The interceptor has already refused the request
// unless it carried the internal secret and an admin role, so this is a label
// for the audit trail rather than the authorisation itself — but it is the label
// that ends up in curated_by, so it is read from the header the gateway sets and
// nowhere else.
func reviewerFromHeaders(h interface{ Get(string) string }) string {
	if v := strings.TrimSpace(h.Get("X-User-Email")); v != "" {
		return strings.ToLower(v)
	}
	return strings.TrimSpace(h.Get("X-User-Id"))
}

File: web/src/app/actions/admin/registerReview.ts
"use server";

// Server actions for the register review console.
//
// EVERY action here calls requireAdmin() ITSELF, not merely because it lives
// under /admin. Server actions are globally addressable by action-id and can be
// POSTed to a route outside the /admin matcher, so the middleware gate alone is
// bypassable — that is precisely the shape of the sendBroadcast defect the
// 2026-07 audit found (a write with no requireAdmin at all).
//
// The RPCs behind these are ALSO in internalOnlyMethods, so the internal secret
// is required on top of an admin identity. A decision here writes a curated
// alias, and a curated alias is the one match method the public gate lets
// publish a live company link against a named politician.

import { createClient } from "@connectrpc/connect";
import { createConnectTransport } from "@connectrpc/connect-web";
import { revalidatePath } from "next/cache";

import { RegisterReviewService } from "~/gen/registerreview/v1/register_review_pb";
import { requireAdmin } from "~/server/admin";
import { SHORTS_API_URL, serverFetchWithUserAgent } from "../config";

/** The decision vocabulary, mirroring the proto enum. No fuzzy member exists. */
export type ReviewDecision =
  | "resolved"
  | "not_a_security"
  | "unlisted_fund"
  | "foreign"
  | "skip";

const DECISION_ENUM: Record<ReviewDecision, number> = {
  resolved: 1,
  not_a_security: 2,
  unlisted_fund: 3,
  foreign: 4,
  skip: 5,
};

export type ReviewAliasKind = "equity" | "etf" | "lic" | "managed_fund";

const ALIAS_KIND_ENUM: Record<ReviewAliasKind, number> = {
  equity: 1,
  etf: 2,
  lic: 3,
  managed_fund: 4,
};

async function reviewClient() {
  const admin = await requireAdmin();
  const transport = createConnectTransport({
    fetch: serverFetchWithUserAgent,
    baseUrl: SHORTS_API_URL,
  });
  return {
    client: createClient(RegisterReviewService, transport),
    headers: {
      "X-Internal-Secret": process.env.INTERNAL_SECRET ?? "dev-internal-secret",
      // curated_by is read from this header server-side, never from the request
      // body: a body-supplied identity is one the caller chose for themselves.
      "X-User-Email": admin.email,
      "X-User-Id": admin.userId,
    },
  };
}

/**
 * Plain-object shapes returned to the client component.
 *
 * The protobuf messages carry BigInt-adjacent generated types and a `$typeName`
 * that does not survive the RSC boundary cleanly, so everything is mapped to
 * plain data here — the same rule the housing surfaces learned (a formatter
 * function or a class instance across the boundary is a runtime error, not a
 * type error).
 */
export interface QueueSample {
  declaredText: string;
  politicianName: string;
  politicianSlug: string;
  itemNo: number;
  parliament: number;
  sourceUrl: string;
}

export interface QueueListing {
  stockCode: string;
  companyName: string;
  existingUses: number;
  isStopwordTicker: boolean;
}

export interface QueueProposal {
  proposedStockCode: string;
  proposedCompanyName: string;
  confidence: number;
  rationale: string;
  model: string;
  status: string;
  shortlist: QueueListing[];
}

export interface QueueItem {
  candidateNorm: string;
  example: string;
  occurrences: number;
  people: number;
  items: number[];
  parliaments: number[];
  entityKinds: string[];
  gateRows: number;
  skipCount: number;
  samples: QueueSample[];
  proposal: QueueProposal | null;
}

export interface Coverage {
  resolved: number;
  listedCandidates: number;
  gatePercent: number;
  backlogCandidates: number;
  backlogRows: number;
  classifiedOut: number;
  method: string;
}

export interface QueuePage {
  items: QueueItem[];
  totalCandidates: number;
  totalRows: number;
}

function toListing(l: {
  stockCode: string;
  companyName: string;
  existingUses: number;
  isStopwordTicker: boolean;
}): QueueListing {
  return {
    stockCode: l.stockCode,
    companyName: l.companyName,
    existingUses: Number(l.existingUses ?? 0),
    isStopwordTicker: !!l.isStopwordTicker,
  };
}

export async function listSecurityQueue(
  limit = 25,
  offset = 0,
  gateOnly = false,
): Promise<QueuePage> {
  const { client, headers } = await reviewClient();
  const resp = await client.listSecurityQueue({ limit, offset, gateOnly }, { headers });

  return {
    totalCandidates: Number(resp.totalCandidates ?? 0),
    totalRows: Number(resp.totalRows ?? 0),
    items: (resp.items ?? []).map((i) => ({
      candidateNorm: i.candidateNorm,
      example: i.example,
      occurrences: Number(i.occurrences ?? 0),
      people: Number(i.people ?? 0),
      items: (i.items ?? []).map(Number),
      parliaments: (i.parliaments ?? []).map(Number),
      entityKinds: i.entityKinds ?? [],
      gateRows: Number(i.gateRows ?? 0),
      skipCount: Number(i.skipCount ?? 0),
      samples: (i.samples ?? []).map((s) => ({
        declaredText: s.declaredText,
        politicianName: s.politicianName,
        politicianSlug: s.politicianSlug,
        itemNo: Number(s.itemNo ?? 0),
        parliament: Number(s.parliament ?? 0),
        sourceUrl: s.sourceUrl,
      })),
      proposal: i.proposal
        ? {
            proposedStockCode: i.proposal.proposedStockCode,
            proposedCompanyName: i.proposal.proposedCompanyName,
            confidence: Number(i.proposal.confidence ?? 0),
            rationale: i.proposal.rationale,
            model: i.proposal.model,
            status: i.proposal.status,
            shortlist: (i.proposal.shortlist ?? []).map(toListing),
          }
        : null,
    })),
  };
}

export async function searchListings(query: string, limit = 15): Promise<QueueListing[]> {
  if (!query.trim()) return [];
  const { client, headers } = await reviewClient();
  const resp = await client.searchListings({ query, limit }, { headers });
  return (resp.listings ?? []).map(toListing);
}

export async function getCoverageStats(): Promise<Coverage | null> {
  const { client, headers } = await reviewClient();
  const resp = await client.getCoverageStats({}, { headers });
  return resp.coverage ? toCoverage(resp.coverage) : null;
}

function toCoverage(c: {
  resolved: number;
  listedCandidates: number;
  gatePercent: number;
  backlogCandidates: number;
  backlogRows: number;
  classifiedOut: number;
  method: string;
}): Coverage {
  return {
    resolved: Number(c.resolved ?? 0),
    listedCandidates: Number(c.listedCandidates ?? 0),
    gatePercent: Number(c.gatePercent ?? 0),
    backlogCandidates: Number(c.backlogCandidates ?? 0),
    backlogRows: Number(c.backlogRows ?? 0),
    classifiedOut: Number(c.classifiedOut ?? 0),
    method: c.method,
  };
}

export interface DecisionResult {
  ok: boolean;
  rowsAffected: number;
  coverage: Coverage | null;
  /** Set when the server refused. Surfaced verbatim — never swallowed. */
  error?: string;
  /** True when the refusal was the stopword-ticker guard, which the UI can retry with a confirmation. */
  needsStopwordConfirm?: boolean;
}

export async function decideSecurityCandidate(input: {
  candidateNorm: string;
  decision: ReviewDecision;
  stockCode?: string;
  aliasKind?: ReviewAliasKind;
  note?: string;
  stopwordConfirmed?: boolean;
}): Promise<DecisionResult> {
  const { client, headers } = await reviewClient();

  try {
    const resp = await client.decideSecurityCandidate(
      {
        candidateNorm: input.candidateNorm,
        decision: DECISION_ENUM[input.decision],
        stockCode: input.stockCode ?? "",
        aliasKind: input.aliasKind ? ALIAS_KIND_ENUM[input.aliasKind] : 0,
        note: input.note ?? "",
        stopwordConfirmed: !!input.stopwordConfirmed,
      },
      { headers },
    );
    // The queue is server-rendered, so the next page load must not serve a
    // candidate this reviewer just decided.
    revalidatePath("/admin/register/securities");
    return {
      ok: true,
      rowsAffected: Number(resp.rowsAffected ?? 0),
      coverage: resp.coverage ? toCoverage(resp.coverage) : null,
    };
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return {
      ok: false,
      rowsAffected: 0,
      coverage: null,
      error: message,
      needsStopwordConfirm: /stopword ticker/i.test(message),
    };
  }
}

export async function undoSecurityDecision(candidateNorm: string): Promise<DecisionResult> {
  const { client, headers } = await reviewClient();
  try {
    const resp = await client.undoSecurityDecision({ candidateNorm }, { headers });
    revalidatePath("/admin/register/securities");
    return {
      ok: !!resp.deleted,
      rowsAffected: 0,
      coverage: resp.coverage ? toCoverage(resp.coverage) : null,
    };
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return { ok: false, rowsAffected: 0, coverage: null, error: message };
  }
}
